File: cpp/2.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <utility>
#include <map>
#include <unordered_map>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2, int add=0){
    if(!l1 && !l2){
      return add ? new ListNode(add) : nullptr;
    }
    else{
      int val = (l1?l1->val:0) + (l2?l2->val:0) + add;
      int nxt_add = 0;
      if(val >= 10)
        val -= 10, ++nxt_add;
      auto t = new ListNode(val);
      t->next = addTwoNumbers(l1?l1->next:nullptr,
                               l2?l2->next:nullptr,
                               nxt_add);
      return t;
    }
  }
};

int main(){
  auto sol = Solution();
  auto l1 = new ListNode(0);
  auto l2 = new ListNode(0);
  auto res = sol.addTwoNumbers(l1, l2);
}

File: python/2.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        add = 0
        ans = []
        while l1 and l2:
            n = l1.val + l2.val + add
            add = n / 10
            n = n % 10
            ans.append(n)
            l1, l2 = l1.next, l2.next
        l = l1 or l2
        while l:
            n = l.val + add
            add = n / 10
            n = n % 10
            ans.append(n)
            l = l.next
        if add:
            ans.append(add)
        h = ListNode(None)
        p = h
        for n in ans:
            p.next = ListNode(n)
            p = p.next
        return h.next
